maven-core/src/main/java/org/apache/maven/plugin/MojoExecutionStrategy.java

```diff
@@ -0,0 +1,46 @@
+package org.apache.maven.plugin;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.lifecycle.LifecycleExecutionException;
+
+import java.util.List;
+
+/**
+ * Interface allows overriding default mojo execution strategy
+ * For example it is possible wrap some mojo execution to decorate default functionality or
+ * skip some executions
+ */
+public interface MojoExecutionStrategy
+{
+
+    void execute( MavenSession session, List<MojoExecution> mojoExecutions, Executor executor )
+            throws LifecycleExecutionException;
+
+    /**
+     * Implementors must perform real mojo execution
+     */
+    interface Executor
```

Comment:
I don't like the idea of subclass, you hardly see it in the Maven codebase. Would it be possible change its usage to `Consumer<MojoExecution>` ?

Reply:
do you mean java.util.function.Consumer? but it since Java 1.8 can we use classes/interfaces from 1.8?

Reply:
regarding subclasses - is there any reason to not use them?
could be moved outside but I'm not sure about naming, the best name here for this interface is MojoExecutor but it is already used

Reply:
Maven 4 will require Java 8, so yes. And by using the Consumer, there's no need for the subclass anymore.

Reply:
ah ok, got it
this branch was created from 3.8
btw would it be possible to merge these changes to 3.8 branch? looks like there is no breaking changes
this would allow us to use/test cache extension immediately and not await and then migrate our project to maven 4

Reply:
My opinion is that we should leave Maven 3 as is and just focus on Maven 4. This feature might attract more people, so putting it into Maven 4 makes more sense to me.

Reply:
> Maven 4 will require Java 8, so yes. And by using the Consumer, there's no need for the subclass anymore.

Robert, what do you think of a bit opposite approach - to refactor and rename MojoExecutor. It is to big and could be split in 3 concerns - project preprocessing, list of mojos processing, idividual mojo execution? MojoExecutor could be renamed to Project(Executor|Builder), and MojoExecutor implements Executor could be extracted

Reply:
> My opinion is that we should leave Maven 3 as is and just focus on Maven 4. This feature might attract more people, so putting it into Maven 4 makes more sense to me.

Robert it make total sense, but from practical needs the extension have to be tested on a stable maven version. Version 4 is in progress and could not be used in real project without upfront investements and also might be not stable yet. So for testing and verification reason need change in version 3 - otherwise there is no way to test extension on a real project

Reply:
Have you tried Maven 4? At its current state I would expect that a Maven project can be built with both Maven 3 and Maven 4. A few warnings have turned into errors, so I would love to see which issues you hit when this feature is merged on the develop branch (aka Maven 4).

Reply:
> My opinion is that we should leave Maven 3 as is and just focus on Maven 4. This feature might attract more people, so putting it into Maven 4 makes more sense to me.

Yes I didn't mean to develop extension in 3.8, I thought just to add these minimal changes to core 3.8 and 4
Then we can use our current extension draft (from 3.8) and test in against real projects (our project is about 750 modules with tricky configuration sometimes)
Our extension built with 3.8 maven could be left as a draft in our fork

To summarise our plan was:

1) merge minimal changes to 3.8 and 4
2) create new branch from MNG-7129-master and apply changes which we have done in our draft extension version (bug fixes, support to change projects versions like it is done by maven set plugin)
3) raise PR from our branch to MNG-7129-master

Reply:
If at all in 3.x I see it in 3.9. I haven't though checked what needs to be changed in Core.

Reply:
> Maven 4 will require Java 8, so yes. And by using the Consumer, there's no need for the subclass anymore.

consumer doesnt declare handled exceptions in api which makes it inconvenient to propagate LifecycleExecutionException. i made executor a functional interface and could extract it to a seprate class if that's preferable